File: Modeling/app/src/main/java/pisarev/com/modeling/mvp/model/DrawVerticalTurning.java
package pisarev.com.modeling.mvp.model;

import android.graphics.Canvas;

import android.graphics.Color;
import pisarev.com.modeling.interfaces.Drawing;
import pisarev.com.modeling.interfaces.IDraw;
import pisarev.com.modeling.mvp.model.base.BaseDraw;

import java.util.List;

public class DrawVerticalTurning extends BaseDraw implements Drawing {

    public DrawVerticalTurning(IDraw draw) {
        super(draw);
    }

    @Override
    public void drawContour(Canvas canvas,MyData data, Point pointCoordinateZero, float zoom, int index) {
        List<Frame> frameList=data.getFrameList();
        boolean isDrawPoint=false;
        Point pStart = new Point();
        Point pEnd = new Point();
        Point point =new Point();
        pStart.setX(650f);
        pStart.setZ(250f);
        pEnd.setX(650f);
        pEnd.setZ(250f);
        float radius;
        float radiusRND;

        for (int i = 0; i < index; i++) {
            if (frameList.get(i).getGCode().contains("G17") || frameList.get(i).getGCode().contains("G18"))
                isG17 = isG17(frameList.get(i).getGCode());
            checkGCode(frameList.get(i).getGCode());
            if (data.getErrorListMap().containsKey(frameList.get(i).getId())) {
                draw.showError(data.getErrorListMap().get(frameList.get(i).getId()));
                break;
            } else {
                if (frameList.get(i).getIsCR() && frameList.get(i).isAxisContains()) {                                  //draw Arc
                    pEnd.setX(frameList.get(i).getX());
                    pEnd.setZ(frameList.get(i).getZ());
                    radius = frameList.get(i).getCr();
                    drawArc(canvas, line, pointCoordinateZero, pStart, pEnd, radius, zoom, clockwise);
                    pStart.setX(pEnd.getX());
                    pStart.setZ(pEnd.getZ());
                }
                if (!frameList.get(i).getIsCR() && !frameList.get(i).isRND() && frameList.get(i).isAxisContains()) {    //draw line
                    pEnd.setX(frameList.get(i).getX());
                    pEnd.setZ(frameList.get(i).getZ());
                    drawLine(canvas, line, pointCoordinateZero, pStart, pEnd, zoom);
                    pStart.setX(pEnd.getX());
                    pStart.setZ(pEnd.getZ());
                }
                if (frameList.get(i).isRND() && frameList.get(i).isAxisContains()) {                                    //draw RND
                    pEnd.setX(frameList.get(i).getX());
                    pEnd.setZ(frameList.get(i).getZ());
                    radiusRND = frameList.get(i).getRnd();
                    Point pointF = new Point();
                    pointF.setX(frameList.get(i + 1).getX());
                    pointF.setZ(frameList.get(i + 1).getZ());
                    drawRND(canvas,line, pointCoordinateZero, pStart, pEnd, pointF, radiusRND, zoom);
                    pStart.setX(pEnd.getX());
                    pStart.setZ(pEnd.getZ());
                    pEnd.setX(frameList.get(i).getX());
                    pEnd.setZ(frameList.get(i).getZ());
                }
                if (isNumberLine && frameList.get(i).getId() == numberLine) {
                    point.setX(frameList.get(i).getX());
                    point.setZ(frameList.get(i).getZ());
                    isDrawPoint=true;
                }
            }
        }
        drawPoint(canvas, pointCoordinateZero, pEnd, colorPoint, zoom);
        if(isDrawPoint) {
            drawPoint(canvas, pointCoordinateZero, point, colorTouchPoint, zoom);
        }
    }

    @Override
    public void setNumberLine(int numberLine) {
        isNumberLine = true;
        this.numberLine=numberLine;
    }
}

File: Modeling/app/src/main/java/pisarev/com/modeling/mvp/contur/DrawVerticalTurning.kt
package pisarev.com.modeling.mvp.contur

import android.graphics.Canvas
import pisarev.com.modeling.interfaces.IDraw
import pisarev.com.modeling.mvp.contur.base.BaseDraw
import pisarev.com.modeling.interfaces.Drawing
import pisarev.com.modeling.mvp.model.core.MyData

class DrawVerticalTurning(draw: IDraw?) : BaseDraw(draw), Drawing {
    override fun drawContour(canvas: Canvas, data: MyData, pointCoordinateZero: Point, zoom: Float, index: Int) {
        val frameList = data.frameList
        var isDrawPoint = false
        val pStart = Point()
        val pEnd = Point()
        val point = Point()

        var radius: Float
        var radiusRND: Float
        for (i in 0 until index) {
            if (frameList[i].gCodes.contains(G17) || frameList[i].gCodes.contains(G18)) isG17 =
                isG17(frameList[i].gCodes)
            checkGCode(frameList[i].gCodes)
            if (data.errorListMap.containsKey(frameList[i].id)) {
                draw!!.showError(data.errorListMap[frameList[i].id]!!)
                break
            } else {
                if (frameList[i].isCR && frameList[i].isAxisContains) {                                  //draw Arc
                    pEnd.x=frameList[i].x.toFloat()
                    pEnd.z=frameList[i].z.toFloat()
                    radius = frameList[i].cr.toFloat()
                    drawArc(canvas,isRapidFeed, pointCoordinateZero, pStart, pEnd, radius, zoom, clockwise)
                    pStart.x=pEnd.x
                    pStart.z=pEnd.z
                }
                if (!frameList[i].isCR && !frameList[i].isRND && frameList[i].isAxisContains) {    //draw line
                    pEnd.x=frameList[i].x.toFloat()
                    pEnd.z=frameList[i].z.toFloat()
                    drawLine(canvas, isRapidFeed, pointCoordinateZero, pStart, pEnd, zoom)
                    pStart.x=pEnd.x
                    pStart.z=pEnd.z
                }
                if (frameList[i].isRND && frameList[i].isAxisContains) {                                    //draw RND
                    pEnd.x=frameList[i].x.toFloat()
                    pEnd.z=frameList[i].z.toFloat()
                    radiusRND = frameList[i].rnd.toFloat()
                    val pointF = Point()
                    pointF.x=frameList[i + 1].x.toFloat()
                    pointF.z=frameList[i + 1].z.toFloat()
                    drawRND(canvas, isRapidFeed, pointCoordinateZero, pStart, pEnd, pointF, radiusRND, zoom)
                    pStart.x=pEnd.x
                    pStart.z=pEnd.z
                    pEnd.x=frameList[i].x.toFloat()
                    pEnd.z=frameList[i].z.toFloat()
                }
                if (isNumberLine && frameList[i].id == numberLine) {
                    point.x=frameList[i].x.toFloat()
                    point.z=frameList[i].z.toFloat()
                    isDrawPoint = true
                }
            }
        }
        drawPoint(canvas, pointCoordinateZero, pEnd, colorPoint, zoom)
        if (isDrawPoint) {
            drawPoint(canvas, pointCoordinateZero, point, colorTouchPoint, zoom)
        }
    }

    override fun setNumberLine(numberLine: Int?) {
        isNumberLine = true
        this.numberLine = numberLine!!
    }


}